Adds support for Kong Admins

tjrivera opened the pull request:
Best reviewed after #18 is merged as this branch contains work included there.

Hi @hbagdi,

This PR is an initial crack at introducing Admin support to `go-kong`. This PR introduces some interesting wrinkles that I think are worth mentioning:

In order to properly test Admin behavior, the test suite has to run against an instance of Kong EE with `enforce_rbac=true` and `admin_gui_auth=basic-auth`.  Additionally, the kong DB is bootstrapped with `KONG_PASSWORD=kong` to introduce a super-admin-user to facilitate testing. Because of this, this PR introduces a `NewTestClient` function that wires up a client with a token found in the `KONG_ADMIN_TOKEN` environment variable to perform authenticated calls against the Admin API. The goal here is to enable the testing of this library against both CE and EE instances. 

It also tweaks the parsing of Kong's semver as reported by Kong. It looks like Kong will report in the form `1.5.0.1-enterprise-edition` for enterprise installations. So this PR strips`-enterprise-edition` and the last patch increment (resulting in this example as `1.5.0`) to fit semver spec. 

Last, I had to update the basic_auth service tests -- It looks like paging behavior in calls to `/basic-auths` likes to include the super-admin user created during bootstrapping which initially broke the test when testing against EE. 

I'm hoping to add a few more methods to Admin when all is said and done (PW Resets, and Role management) but figured this would be a good place to stop and submit for feedback. 

Thanks!

- Tyler

tjrivera commented:
Addresses #17

hbagdi commented:
Please see #23 and let me know your thoughts.
This PR also includes some changes from #18, please remove those.

tjrivera commented:
Hey @hbagdi, thanks for the review. I like #23 and am happy to incorporate that here once it's merged. 

I'm having some trouble deciding on how to structure these PRs since the testing of the `Admin` entity code here relies on functionality built out in the workspaces PR. Similarly, a lot of the RBAC stuff relies on workspaces. Any suggestions are appreciated. 

Thanks,
Tyler

Commit:
feat: adds kong admins support

Adds support for Kong admins

* Introduces NewTestClient which allows for authenticated calls to the admin API based on the presence of a KONG_ADMIN_TOKEN env var
* Adds basic Admin CRUD functionality
* Includes a "fix" to semantic versioning used on Kong EE releases

Moves HeaderRoundTripper

Addresses golangci-lint issues

hbagdi commented:
> I'm having some trouble deciding on how to structure these PRs since the testing of the Admin entity code here relies on functionality built out in the workspaces PR. Similarly, a lot of the RBAC stuff relies on workspaces. Any suggestions are appreciated.

We will have to take it one a time for such PRs.
You can open up multiple PRs with history of one PR included in another PR. Then, you can rebase on top of master as other PRs get merged in. It will lead to a little bit of extra work but that's important to keep the code quality up to the standards we need.

Commit:
chore(ci) enable rbac for kong ee

Commit:
chore: remove workspace entities service

Commit:
chore: add test client to mtls, upstream, and workspace

Commit:
chore: add build tag. adds rbac flag for ee runWhen

tjrivera commented:
Hey @hbagdi, Sorry for the delay in progress here. I've been pretty swamped with some of our internal releases and am excited to tidy up some of the work I've been doing on those efforts so that I can contribute back. 

I rebased this branch against the main branch and removed the workspace entities work. 

You'll notice that I've added a `NewTestClient` function whose main purpose is to return a client that injects the `kong-admin-token` header if `KONG_ADMIN_TOKEN` is present as an environment variable. I realized that a lot fo the upcoming functionality relies on rbac being enabled to test properly, so I've also updated the ci workflow -- though it looks like something here is still broken -- I don't think we have the enterprise debian packages in our bintray so I couldn't dig much deeper.

I also added a `rbac` argument to the `runWhenEnterprise` function to be able to indicate if a test requires rbac or not. 

I have the rest of EE functionality built out, I just have to take some time to package them up into distinct PRs, so be on the lookout for those. 

Thanks for your time,
Tyler

hbagdi commented on kong/types.go:
```diff
@@ -220,3 +220,26 @@ type Workspace struct {
+	UniqueFieldValue *string `json:"unique_field_value,omitempty" yaml:"unique_field_value,omitempty"`
+	WorkspaceID      *string `json:"workspace_id,omitempty" yaml:"workspace_id,omitempty"`
+	WorkspaceName    *string `json:"workspace_name,omitempty" yaml:"workspace_name,omitempty"`
+}
```
Please remove this unrelated change from the PR.

hbagdi commented on kong/kong_test.go:
```diff
@@ -162,3 +173,40 @@ func TestRunWhenEnterprise(T *testing.T) {
+}
+
+// RoundTrip satisfies the RoundTripper interface.
+func (t *HeaderRoundTripper) RoundTrip(req *http.Request) (*http.Response,
```
We already implement a HeaderRoundTripper:
https://github.com/hbagdi/go-kong/blob/master/kong/utils.go#L79

Please re-use this one.

hbagdi commented on kong/admin_service_test.go:
```diff
@@ -0,0 +1,180 @@
+}
+
+func TestAdminServiceWorkspace(T *testing.T) {
+	runWhenEnterprise(T, ">=0.33.0", false)
```
These tests with `rbac=false` won't be executed in the current CI pipeline. Does this mean that we need to include another matrix parameter in the CI to run all tests with combinations of rbac=true and false?

What is kong's behavior here if you run this test with rbac=off in kong?

tjrivera replied:
Kong will throw 500s if a request is made to an endpoint that utilizes RBAC entities with `rbac=off`. `TestAdminServiceRegisterCredentials` is an example of this.

I believe these tests will run in the current pipeline. When the `rbac` parameter in `runWhenEnterprise` is true it signals that this test needs RBAC enabled to run properly. I think that the EE tests in the CI pipeline should be run with RBAC enabled by default. I don't see a downside to that. I included this additional parameter in case folks needed to run the test suite against an instance of Kong EE that for whatever reason didn't have RBAC enabled. 

I'm going to rename the function parameter for more clarity

Commit:
chore: remove unused type definition

Commit:
chore: reuse headerRoundTripper

* also renames runWhenEnterprise parameter from rbac to rbacRequired to add clarity to use case.
* changes header definitions for NewTestClient based on headerRoundTripper refactor

tjrivera commented:
Hi @hbagdi, 

I spent a little time this morning getting access to the repo that GitHub Actions is using here to try and get EE CI Tests working. I was able to confirm that the setup script you have here _should_ work. The only thing that could be an issue is the credentials that are being passed, but that's not something I have insight on. In my own testing I noticed that bintray wants the API key associated with the user instead of the password when passing creds via basic auth -- not sure if that makes a difference here. I'd love to see those tests pass!

hbagdi commented on kong/types.go:
```diff
@@ -220,3 +220,16 @@ type Workspace struct {
+	CreatedAt        *int    `json:"created_at,omitempty" yaml:"created_at,omitempty"`
+	ID               *string `json:"id,omitempty" yaml:"id,omitempty"`
+	Email            *string `json:"email,omitempty" yaml:"email,omitempty"`
+	Password         *string `json:"password,omitempty" yaml:"password,omitempty"`
```
I don't see this field being present in the API response on Kong Enterprise.

tjrivera replied:
I feel like I've seen this in the wild but also can't find anything to support that in the docs or tests against an RBAC-enabled instance. I'll remove

hbagdi commented on kong/admin_service.go:
```diff
@@ -0,0 +1,174 @@
+
+	var createdAdmin struct {
+		Admin Admin `json:"admin,omitempty" yaml:"admin,omitempty"`
+	}
```
This bit seems like a hack in the API but the smell is Kong Enterprise and not this patch. Leaving a note for future reference, no changes required.

hbagdi commented on kong/admin_service.go:
```diff
@@ -0,0 +1,174 @@
+	admin *Admin) error {
+
+	if admin == nil {
+		return errors.New("cannot register credentials on a nil Admin")
```
credentials _for_ a nil Admin

hbagdi commented on kong/admin_service.go:
```diff
@@ -0,0 +1,174 @@
+
+// Get fetches a Admin in Kong.
+func (s *AdminService) Get(ctx context.Context,
+	nameOrID *string, generateRegisterURL bool) (*Admin, error) {
```
Can we split this into two functions?
Get(ctx, string): This one fetches the admin
GenerateRegisterURL(ctx,string): This one generates a register URL.

Commit:
fix: typos, refactor admin retrieval, clean up admin type

Commit:
fix: adding the password field back

the password field is necessary to register credentials (Ref: `AdminService.RegisterCredentials`)

hbagdi commented:
Here is the CI run:
https://github.com/hbagdi/go-kong/runs/812274288

The tests pass for this branch but the merge tests don't pass.

hbagdi commented:
Can you rebase on master/main? I merged in #28 and that would provide us some insight as to what is going on.

(I'm renaming the default branch from master to main, hence the two branches)

hbagdi commented:
Merged this manually. Great work!
